Lib/typing.py

```diff
@@ -2884,8 +2884,12 @@ def __new__(cls, name, bases, ns, total=True):
 
         for base in bases:
             annotations.update(base.__dict__.get('__annotations__', {}))
-            required_keys.update(base.__dict__.get('__required_keys__', ()))
-            optional_keys.update(base.__dict__.get('__optional_keys__', ()))
+            base_required = base.__dict__.get('__required_keys__', ())
+            base_optional = base.__dict__.get('__optional_keys__', ())
+            required_keys.update(base_required)
+            required_keys.difference_update(base_optional)
+            optional_keys.update(base_optional)
+            optional_keys.difference_update(base_required)
```

Comment:
It took me a minute or two to wrap my head around this. I'd find this a little more readable:

```suggestion

            base_required = base.__dict__.get('__required_keys__', set())
            required_keys |= base_required
            optional_keys -= base_required

            base_optional = base.__dict__.get('__optional_keys__', set())
            required_keys -= base_optional
            optional_keys |= base_optional
```